### ISSUE TITLE
Structural comparison vs. performance of oneWayLazy 

### ISSUE THREAD

**Issue by cmeeren** [elmish/Elmish.WPF#49]
Consider the following lines used for updating a `oneWayLazy` binding and indicating if `PropertyChanged` should be triggered (this is from the rewrite, but it works similarly in the old implementation):

https://github.com/elmish/Elmish.WPF/blob/89b62f43a76eb50ccf56e68e20448b4557254607/src/Elmish.WPF/ViewModel.fs#L155-L160

Half the point of this `oneWayLazy` is to avoid running `map` unless its input (i.e., the output from `get`) has changed, so there must be some kind of equality check here. However, as it's currently implemented, it compares using `=` which means that we might potentially be structurally comparing, say, large lists of records (certainly not a far-fetched assumption considering the usecase of `oneWayLazy`). AFAIK this can be an expensive operation itself, and this would be called for absolutely all updates for this particular (sub)model.

**I'm wondering if anyone has any tips on how to improve this performance hit.**

As can be seen from the code comment, I originally thought about using reference equality, but this won't work because if `get` extracts several items from the state and returns them in a tuple (or any other type, for that matter), `Object.ReferenceEquals` will return `false`.

(There are also other problems with reference equality, such as change detection potentially yielding false positives for different objects that are structurally identical, or false negatives for if `get` returns a mutable object, but those aren't as serious if one has an idiomatic, immutable model. I originally wrote a bit about those two issues here, but I then realized that reference equality wasn't feasible at all due to the aforementioned problem, so I'll spare you the details.)

**Comment by cmeeren** [comment]
A very simple workaround/method (on the library side) is to add an `isEqual : 'a -> 'a -> bool` parameter to `Binding.oneWayLazy`, allowing users to optimize as needed. It could be as simple as `(=)` (which is the current behavior), but it could also be e.g.

```f#
fun (simple1, list1) (simple2, list2) ->
  simple1 = simple2 && Object.ReferenceEquals(list1, list2)
```

This should of course be documented to provide clear guidance.

**Comment by cmeeren** [comment]
Yes, I like that very much. Thanks! 👍

**Comment by cmeeren** [comment]
Closed by a5e11703605ca1cd2c14d5353259051179f1685b
